runtime/vm/OutOfLineINL_openj9_internal_foreign_abi_InternalDowncallHandler.cpp

```diff
@@ -216,6 +224,61 @@ OutOfLineINL_openj9_internal_foreign_abi_InternalDowncallHandler_initCifNativeTh
 	ffiTypeHelpers.freeStructFFIType(returnType);
 	goto done;
 }
+
+#if JAVA_SPEC_VERSION >= 22
+/* openj9.internal.foreign.abi.InternalDowncallHandler: private static native boolean isFfiProtoEnabled(); */
+VM_BytecodeAction
+OutOfLineINL_openj9_internal_foreign_abi_InternalDowncallHandler_isFfiProtoEnabled(J9VMThread *currentThread, J9Method *method)
+{
+	VM_BytecodeAction rc = EXECUTE_BYTECODE;
+	J9JavaVM *vm = currentThread->javaVM;
+	jboolean isFfiProtoOn = J9_ARE_ANY_BITS_SET(vm->extendedRuntimeFlags2, J9_EXTENDED_RUNTIME2_FFI_PROTO) ? JNI_TRUE : JNI_FALSE;
+
+	VM_OutOfLineINL_Helpers::returnSingle(currentThread, isFfiProtoOn, 0);
+	return rc;
+}
+
+/**
+ * openj9.internal.foreign.abi.InternalDowncallHandler: private native void resolveBoundHandle(MethodHandle handle);
+ *
+ * @brief Resolve the bounded handle via MethodHandleResolver.ffiCallLinkCallerMethod() in advance
+ *        when calling into the interpreter via MH.linkToNative()
+ *
+ * @return void
+ */
+VM_BytecodeAction
+OutOfLineINL_openj9_internal_foreign_abi_InternalDowncallHandler_resolveBoundHandle(J9VMThread *currentThread, J9Method *method)
+{
+	VM_BytecodeAction rc = EXECUTE_BYTECODE;
+	j9object_t invokeCache = NULL;
+
+	jobject nativeMH = j9jni_createLocalRef((JNIEnv *)currentThread, J9_JNI_UNWRAP_REFERENCE(currentThread->sp));
```

Comment:
My apologies, there is a problem with this also. Local refs require a jni frame, which you dont have in this case.

I still prefer to not use global refs since that is more heavy weight.

The other option is to use `VM_VMHelpers::push/popObjectInSpecialFrame(currentThread, srcObject);` but you need to do that after the `buildInternalNativeStackFrame` has been done.

Reply:
Yeah, Technically `push/popObjectInSpecialFrame1` should be ideal to handle our case across the GC point. Let me try this out to see how it goes.

Reply:
Replaced the local refs with `VM_VMHelpers::push/popObjectInSpecialFrame` which works good in tests as expected.